### ISSUE TITLE
Migrate to yarn

### ISSUE THREAD

**Issue by scinos** [Automattic/wp-calypso#40882]
This issue is used to keep track of the migration from npm to yarn.

# Summary
We move wp-calypso repo from npm to yarn (work already started by @blowery and @jameslnewell  in #39121).

The biggest risk with this transition is accidentally altering the dependency tree. This could introduce production bugs at the same time we are changing our tools, making everything extra harder to diagnose and fix. 

To ensure a clean transition, we first need to have a clean `npm` dependency tree.

# Preparation

These can be merged to master

- [x] Develop a tool for computing the effective tree (https://www.npmjs.com/package/@automattic/effective-module-tree)
- [x] Update @automattic/color-studio to 2.2.1 (#40167)
- [x] Adds missing dependencies (#40165, #40057)
- [x] Link all sub-packages in main `package.json` (#39955)
- [x] Move `devDependencies` to `dependencies` (workaround for yarnpkg/yarn#7807) (#40880)
- [x] Change scripts to ensure they are run with the correct package manager (assert `npm` is used if `package-lock.json` is present, `yarn` if `yarn.lock` is present) (#40893)
- [x] Make dependency of node-sass explicit (#41073)

# Migration to yarn

To keep things easy to review, we'll have a long-run feature branch `update/npm-to-yarn-main`. All these changes will be merged into that branch.

- [x] Change documentation from npm to yarn (#40916)
- [x] Use yarn policy to set a version (#40902)
- [x] Change all scripts to use `yarn` commands (#40921)
- [x] Enable yarn workspaces (#40923)

Once all the above have been reviewed and merged to`update/npm-to-yarn-main`, we'll do a final merge to master.

**Comment by scinos** [comment]
@nerrad @simison @nb @blowery @dmsnell @tyxla @sirbrillig @zinigor @sirreal @griffbrad @p-jackson @ChaosExAnima @brbrr @stojdza @senadir @creativecoder 

Hey folks! In the past, you have shown some interest in Calypso migrating to `yarn`. In case you are still interested, this is the issue where I'm capturing the state of the project. Please, feel free to check the open PRs and the overall plan and let me know if you have any feedback or directly comment on the relevant PRs.

**Comment by nb** [comment]
Thanks for all the work on that, @scinos (and before @blowery and @jameslnewell)! This is exciting, looking forward to eliminating a whole class of tricky problems.

A couple of (hopefully not too high-level) questions:

* What is our criteria for being ready to complete the migration? How do we know it's stable enough? Do we expect that the `nom` and `yarn` trees are the same?
* What are ways in which we can reduce the risk for production bugs once we deploy?

**Comment by scinos** [comment]
@nb There are two main criteria for the migration:

1. The changes in the tree must be contained, small, expected and explainable). We know `npm` can generate a tree that can't be represented in `yarn` with 100% fidelity, so there will be some changes.

2. Calypso bundles must not change significatively when comparing `yarn` vs `npm`. As above, there might be some minor changes.

I think that if we can guarantee the above two points, we will be in a good spot and the migration won't introduce any new bug.

**Comment by nb** [comment]
Thanks, @scinos.

“Contained, small, expected and explainable” change to the tree makes perfect sense.

As for the bundles, I am curious to see whether we will be able to reason about the changes we see.

Do you think it's worth trying to test on a smaller audience? Like wordpress.com staging (seen by proxied Automatticians)? Or it will be too much effort?

**Comment by scinos** [comment]
For the record, the complete list of changes in the tree is listed in https://github.com/Automattic/wp-calypso/pull/40923#issue-400906644. The changes in the bundles are listed in https://github.com/Automattic/wp-calypso/pull/40923#issuecomment-614028212

At this point, I think we are pretty safe. The only bit that slightly concerns me is the duplication of `acorn` (see comment in the second link) because it adds 27kb to 3 async chunks. I assume those chunks are only loaded in some specific use cases, so most users won't be impacted. At any rate, deduplicating `acorn` to migitate it should be pretty high in our priority list once yarn lands in master.

As for testing in a subset of users... in this specific case, I think it is too much effort. Keeping `npm` and `yarn` alive for a few days/weeks means we have to keep `yarn` in sync with every change in the dependency tree in `npm`. That process is partially manual, and therefore error prone. Which means we can introduce _more_ risks, but specially, it means that the day we press the button to merge it "for good", we will potentially be merging a dep tree different than the tree we would have tested in staging.

**Comment by scinos** [comment]
This project has been merged in master. The final list of dependency changes are listed in https://github.com/Automattic/wp-calypso/pull/41140#issuecomment-614592351.
